arch/x86/include/asm/kvm_pkvm.h

```diff
@@ -200,6 +200,29 @@ static inline unsigned long pkvm_shadow_ept_pgtable_pages(int nr_vm)
 	return (res * 2);
 }
 
+/*
+ * Calculate the total pages for shadow IO page tables for Legacy IOMMU.
+ * Similarly to the calculation for shadow EPT, we assume that there is
+ * no shared memory between devices using different page tables.
+ */
+static inline unsigned long pkvm_shadow_iommu_pgtable_pages(int nr_pdev)
+{
+	unsigned long pgtable_pages = __pkvm_pgtable_total_pages();
+	unsigned long res;
+
+	res = pgtable_pages;
+
+	/*
+	 * Similarly to shadow VMs (see the comment in
+	 * pkvm_shadow_ept_pgtable_pages()), each device may require
+	 * its own level2:level5 page table pages.
+	 */
+	res += __pkvm_pgtable_max_pages(pgtable_pages) * (nr_pdev - 1);
```

Comment:
Is this enough or maybe we need twice as many pages to support vSMM mode like in `pkvm_shadow_ept_pgtable_pages()`? (I know nothing about vSMM.)

Reply:
> I don't think vSMM will use pass-thru devices.

Cool.

> BTW, here the shadow iommu pgtable is actually for host VM, could it be pkvm_host_shadow_iommu_pgtable_pages?
(I have to say pgstate table is also a shadow iommu pgtable, but it's per L2 VM. Here shadow iommu pgtable is specific for host VM.)

To be precise, shadow iommu pgtable may be also used for normal VM. But normal VM devices are still "owned" by the host VM (not attached to any L2 VM), so I'm ok with renaming this to `pkvm_host_shadow_iommu_pgtable_pages`.

> And could we make it be 0 if IOMMU is scalable mode?

That would be nice (to save memory) but the problem is that, as I understand, this calculation is used by the host in `pkvm_reserve()` which is done very early, before we know anything about IOMMUs.
We can add a Kconfig option for disable legacy IOMMU support, OTOH it probably would be good to avoid adding compilation options for things that can be detected automatically.
So I was assuming that we can reserve this memory unconditionally for now, and in the future, if needed, think how to disable reserving it if there is no legacy IOMMU.

Reply:
> > BTW, here the shadow iommu pgtable is actually for host VM, could it be pkvm_host_shadow_iommu_pgtable_pages?
> > (I have to say pgstate table is also a shadow iommu pgtable, but it's per L2 VM. Here shadow iommu pgtable is specific for host VM.)
> 
> To be precise, shadow iommu pgtable may be also used for normal VM. But normal VM devices are still "owned" by the host VM (not attached to any L2 VM), so I'm ok with renaming this to `pkvm_host_shadow_iommu_pgtable_pages`.
> 

Yes, normal VM's devices are still owned by the host. But in our design, for normal VM, we are using pgstate table as the shadow iommu pgtable. It's different with this one.

> > And could we make it be 0 if IOMMU is scalable mode?
> 
> That would be nice (to save memory) but the problem is that, as I understand, this calculation is used by the host in `pkvm_reserve()` which is done very early, before we know anything about IOMMUs. We can add a Kconfig option for disable legacy IOMMU support, OTOH it probably would be good to avoid adding compilation options for things that can be detected automatically. So I was assuming that we can reserve this memory unconditionally for now, and in the future, if needed, think how to disable reserving it if there is no legacy IOMMU.

Got it, I am okay for it.

Reply:
> But in our design, for normal VM, we are using pgstate table as the shadow iommu pgtable. It's different with this one.

Hmm, don't we use pgstate table as the iommu pgtable for protected VMs only? ptdev->pgt is set to pgstate_pgt only in pkvm_attach_ptdev(), which is called only for protected VM. Maybe I'm misunderstanding your point.

Reply:
> > But in our design, for normal VM, we are using pgstate table as the shadow iommu pgtable. It's different with this one.
> 
> Hmm, don't we use pgstate table as the iommu pgtable for protected VMs only? ptdev->pgt is set to pgstate_pgt only in pkvm_attach_ptdev(), which is called only for protected VM. Maybe I'm misunderstanding your point.

For normal VM, we always use pgstate table to maintain page state.
And for normal VM under scalable IOMMU, we use host ept as IOMMU pgtable.
While for normal VM under legacy IOMMU, yes, we can use IOMMU spgt as IOMMU pgtable, but I think this spgt should combine with pgstate pgtable. Otherwise, we will have both IOMMU spgt and pgstate pgt.

Reply:
I think currently we may keep both IOMMU spgt & pgstate pgt for normal VM under legacy IOMMU mode, it will not make your patch complicated as now you don't have page state control for IOMMU spgt at all. We can consider how to add page state control for all IOMMU spgt (include normal VM one) in the future.

Reply:
- Renamed to `pkvm_host_shadow_iommu_pgtable_pages`
- Added TODO for not reserving memory if there is no legacy IOMMU